bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/device/PJLinkDevice.java

```diff
@@ -0,0 +1,314 @@
+        Instant now = Instant.now();
+        boolean connectionTooOld = false;
+        if (this.socketCreatedOn != null) {
+            long milliseondsSinceLastConnect = Duration.between(this.socketCreatedOn, now).toMillis();
+            // according to the PJLink specification, the device closes the connection after 30s idle (without notice),
+            // so to be on the safe side we do not reuse sockets older than 20s
+            connectionTooOld = milliseondsSinceLastConnect > 20 * 1000;
+        }
+
+        if (forceReconnect || connectionTooOld) {
+            if (socket != null) {
+                socket.close();
+                socket = null;
+            }
+        }
+
+        this.socketCreatedOn = now;
+        if (socket != null && socket.isConnected() && !socket.isClosed()) {
+            return;
+        }
+
+        SocketAddress socketAddress = new InetSocketAddress(ipAddress, tcpPort);
+
+        try {
+            this.socket = new Socket();
+            socket.connect(socketAddress, timeout);
+            socket.setSoTimeout(timeout);
+            this.reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+            String header = reader.readLine();
+            if (header == null) {
+                throw new ResponseException("No PJLink header received from the device");
+            }
+            header = header.toUpperCase();
+            switch (header.substring(0, "PJLINK x".length())) {
+                case "PJLINK 0":
+                    logger.info("Authentication not needed");
+                    this.authenticationRequired = false;
+                    break;
+                case "PJLINK 1":
+                    logger.warn("Authentication needed");
+                    this.authenticationRequired = true;
+                    if (this.adminPassword == null) {
+                        this.socket.close();
+                        throw new AuthenticationException("No password provided, but device requires authentication");
+                    } else {
+                        try {
+                            this.authenticate(header.substring("PJLINK 1 ".length()));
+                        } catch (AuthenticationException e) {
+                            // propagate AuthenticationException
+                            throw e;
+                        } catch (ResponseException e) {
+                            // maybe only the test command is broken on the device
+                            // as long as we don't get an AuthenticationException, we'll just ignore it for now
+                        }
+                    }
+                    break;
+                default:
+                    logger.warn("Cannot handle introduction response {}", header);
+                    throw new ResponseException("Invalid header: " + header);
+            }
+        } catch (ConnectException | SocketTimeoutException | NoRouteToHostException e) {
+            throw e;
```

Comment:
those are subclasses of java.io.IOException (which is declared), this should be enough right?

Reply:
I think you're right.
But why catching them to then only throw them ?

Reply:
The first catch is more specific (second catch has IOException), as these Exceptions indicate that there is just no device to connect to (during discovery). I wanted to avoid flooding the log with this information during scan. Except for these three subclasses, IOExceptions seemed to be interesting enough to write a log entry (second catch)